chore: adding more required functionality

priteshbandi opened the pull request:
* Adding more functionality required for notation-go to depend on this pkg.
* Also some renaming and update to Error struct to matching it with notation-go.

Two-Hearts reviewed (commented):
1. Please add doc to struct `CLI` defined at line 23.
2. When I new a `CLI struct` with `pl == nil`: `cli, _ := New("notation-my-plugin", nil)`, this will `panic` when I call `cli.Execute(...)`. We should always check `nil` when new a `CLI`.
3. In `NewWithLogger` function, we are only checking prefix of the plugin executable name. We should also check the suffix matches with the plugin's `metadata.Name`, i.e. `{plugin-name}` of `notation-{plugin-name}` should equal to `pl.GetMetadata resp.Name`. In fact, do we really need the `executableName` parameter here?
4. `deferStdout()` and `type discardLogger struct{}` should be moved to `utils.go`
5. In `utils.go getValidArgs()`, the result strings should not be hardcoded, because they are defined in the `proto.go`: https://github.com/notaryproject/notation-plugin-framework-go/blob/c8c68d2fe13bd094d1bed57886d78a10d3ba5618/plugin/proto.go#L27

shizhMSFT commented on plugin/proto.go:
```diff
@@ -26,6 +32,11 @@ const (
 // Command is a CLI command available in the plugin contract.
 type Command string
 
+// Request defines a plugin request, which is always associated to a command.
+type Request interface {
+	Command() Command
+}
```
What's the purpose of this interface?

priteshbandi replied:
Well, technically we shouldn't need this but the way its currently implemented in [notation-go](https://github.com/search?q=repo%3Anotaryproject%2Fnotation-go%20Command&type=code) we need this interface.

priteshbandi commented:
> 1. Please add doc to struct `CLI` defined at line 23 of `cli.go`.
> 
>     2. When I new a `CLI struct` with `pl == nil`: `cli, _ := New("notation-my-plugin", nil)`, this will `panic` when I call `cli.Execute(...)`. We should always check `nil` when new a `CLI`.
> 
>     3. In `NewWithLogger` function, we are only checking prefix of the plugin executable name. We should also check the suffix matches with the plugin's `metadata.Name`, i.e. `{plugin-name}` of `notation-{plugin-name}` should equal to `pl.GetMetadata resp.Name`. In fact, do we really need the `executableName` parameter here?
> 
>     4. `deferStdout()` and `type discardLogger struct{}` should be moved to `utils.go`
> 
>     5. In `utils.go getValidArgs()`, the result strings should not be hardcoded, because they are defined in the `proto.go`: https://github.com/notaryproject/notation-plugin-framework-go/blob/c8c68d2fe13bd094d1bed57886d78a10d3ba5618/plugin/proto.go#L27

Done. Addressed all the feedbacks

priteshbandi commented on plugin/proto.go:
```diff
@@ -26,6 +32,11 @@ const (
 // Command is a CLI command available in the plugin contract.
 type Command string
 
+// Request defines a plugin request, which is always associated to a command.
+type Request interface {
+	Command() Command
+}
+
```
tracking it in https://github.com/notaryproject/notation-plugin-framework-go/pull/5#discussion_r1436200680

Two-Hearts commented on cli/cli.go:
```diff
@@ -19,24 +20,28 @@ import (
-func NewWithLogger(executableName string, pl plugin.Plugin, l log.Logger) *CLI {
+func NewWithLogger(pl plugin.Plugin, l log.Logger) (*CLI, error) {
+	if pl == nil {
+		return nil, errors.New("plugin cannot be nil")
+	}
+
 	return &CLI{
-		name:   executableName,
 		pl:     pl,
 		logger: l,
-	}
+	}, nil
```
We have the `name` field defined in `CLI` struct. In `NewWithLogger`, we should get the plugin's metadata and fill `CLI.name` with `md.Name`.

priteshbandi replied:
Removed the field as its unused.

Two-Hearts reviewed (commented):
LGTM with a single nit comment.

JeyJeyGao commented on cli/utils.go:
```diff
@@ -16,27 +17,89 @@ func getValidArgsString(md *plugin.GetMetadataResponse) string {
-	sort.Strings(opts)
-	return opts
+	// convert &opts (type *[]Command) to *[]string via unsafe.Pointer, then deref
+	stringOpts := *(*[]string)(unsafe.Pointer(&opts))
```
Can we avoid using unsafe.Pointer here?
For example:
```go
opts := []string{
		string(plugin.CommandGetMetadata), string(plugin.Version),
	}
```

priteshbandi replied:
Yep that's possible but reduces readability and performance. Is ther4e any specific concern with  unsafe.Pointer as command is basically an alias for string?

JeyJeyGao commented on cli/cli.go:
```diff
@@ -19,24 +20,28 @@ import (
 
+// CLI struct is used to create an executable for plugin.
 type CLI struct {
 	name   string
```
Do we need the name field?

priteshbandi replied:
Removed

Two-Hearts reviewed (approved):
LGTM

rgnote reviewed (approved):
LGTM

JeyJeyGao reviewed (approved):
LGTM
